specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,484 @@
+and results consecutively such that there are no unused values within these
+ranges.
+
+### Qubit and Result Usage
+
+The amount of available memory on a QPU both with regards to qubits and
+potentially also with regards to classical memory for storing measurement
+results before they are read out and transmitted to another classical processor
+is commonly still fairly limited. Any memory - quantum or classical - that is
+used during quantum execution is not usually managed dynamically. Instead,
+operations are scheduled and resources are bound as part of compilation. How
+early in the process this happens varies, and QIR permits to express programs in
+a form that either defers allocation and management of such resources to later
+stages, or to directly identify individual qubits and results by a constant
+integer value as outlined above. This permits to accurately reflect application
+intent for a variety of frontends.
+
+Ultimately, it is up to the executing backend what data structure is associated
+with a qubit or result value. This gives a backend the freedom to, e.g., process
+measurement results asynchronously, or attach additional device data to qubits.
+Qubit and result values are correspondingly represented as opaque pointers in
+the bitcode, and a QIR program must not dereference such pointers, independent
+on whether they are merely bitcasts of integer constants as they are in the Base
+Profile program above, or whether they are created dynamically, meaning the
+value is managed by the executing backend.
+
+To execute a given bitcode, the backend needs to know how to process qubit and
+result pointers used by a program. This is achieved by storing metadata in the
+form of [module flags](#module-flags-metadata) in the bitcode. QIR does not make
+a type distinction or for example uses a different address space for the two
+kinds of pointers. This ensures that libraries and optimization passes that map
+between different instruction sets do not need to distinguish whether the
+compiled application code makes use of dynamic qubit and result management or
+not.
+
+To be compliant with the Base Profile specification, the program must not make
+use of dynamic qubit or result management; instead, qubits and results must be
+identified by a constant integer value that is bitcast to a pointer to match the
+expected type. How such an integer value is interpreted and specifically how it
+relates to hardware resources is ultimately up to the executing backend.
+
+Additionally, the Base Profile imposes the following restrictions on qubit and
+result usage:
+
+- Qubits must not be used after they have been measured. Which quantum
+  instructions perform measurements is defined in the specification for known
+  [quantum instructions](../Instruction_Set.md). For quantum instructions that
+  are not listed there, we refer to the documentation provided by the targeted
+  backend that supports them.
```

Comment:
This is going to make it tricky for QAT to validate this rule universally...

Reply:
I think sooner or later we need a more formal definition regarding what classifies as what. However, I also think we need a bit more experience value before we can do that. Instructions that are supported by a hardware backend should be listed in the Instruction_Set.md. This sentence regarding backend documentation is primarily intended for research tools (given how fast paced and diverse the needs are for those, I believe it is good to allow room for having all sorts of "out of protocol" experimental features there).

Reply:
After some thinking, I did end up editing this in favor of introducing what I hope is a good choice also going forward for other profiles: The spec now requires to add an attribute to QIS functions that are not reversible, and specifically to the functions that measure qubits. It is up to the authors of the target package that defines a QIS to tag these declarations appropriately, so in principle, what is or isn't reversible could be backend dependent. I am not sure whether the full spec should have specific requirements in that regard for the functions that are listed as "known quantum instructions", but it certainly defines a clear mechanism in the IR to deal with functions that are "outside the official spec" in an automated and clearly defined manner.